python/tk_multi_launchapp/software_entity_launcher.py

```diff
@@ -0,0 +1,196 @@
+# Copyright (c) 2016 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+import pprint
+
+import sgtk
+
+from .base_launcher import BaseLauncher
+
+shotgun_data = sgtk.platform.import_framework("tk-framework-shotgunutils", "shotgun_data")
+
+class SoftwareEntityLauncher(BaseLauncher):
+    """
+    Launches a DCC application based on site Software entity entries.
+    """
+    def register_launch_commands(self):
+        """
+        Determine what launch command(s) to register with the current TK engine.
+        Multiple commands may be registered based on the number of retrieved
+        Software entities and their corresponding 'versions' field.
+        """
+        # Determine the information to retrieve from Shotgun
+        # @todo: The 'sg_software_entity' setting can be removed once the
+        #        Software entity becomes native.
+        sw_entity = self._tk_app.get_setting("sg_software_entity") or "Software"
+
+        # Use filters to retrieve Software entities that match specified
```

Comment:
I think this procedural approach makes the logic hard to read. For me, i feel it's clearer with something like this:

```python
        filters = [{

            # AND together all of the following filter conditions
            "filter_operator": "and",

            "filters": [

                # make sure to only include active entries
                ["sg_status_list", "is", "act"],

                # handle project constraints -
                # only get entries where either the project field
                # is unspecified or when it includes the current project
                {
                    "filter_operator": "or",
                    "filters": [
                        ["sg_projects", "is", None],
                        ["sg_projects", "in", current_project]
                    ]
                },

                # handle user and group constraints -
                # get fields where either
                # both the user and group field is unspecified
                # (means no restrictions for this launch record)
                # or
                # where the current user is either contained in the list of users
                # or in the list of groups
                {
                    "filter_operator": "or",
                    "filters": [
                        {
                            # both group restrictions and user restrictions
                            # are empty,
                            "filter_operator": "and",
                            "filters": [
                                ["sg_user_restrictions", "is", None],
                                ["sg_group_restrictions", "is", None]
                            ]
                        },
                        {
                            # current_user is either in the user restrictions field or  
                            # part of a group in the group restrictions field.
                            "filter_operator": "or",
                            "filters": [
                                ["sg_user_restrictions", "in", current_user],
                                ["sg_group_restrictions.Group.users", "in", current_user]
                            ]
                        },
                    ]
                },
            ]
        }]
```